vendors/st/boards/stm32l475_discovery/ports/common_io/iot_uart.c

```diff
@@ -0,0 +1,689 @@
+        }
+        else
+        {
+            HAL_NVIC_SetPriority( ( IRQn_Type ) ( ( STM32_UART_HalContext_t * ) pxUartPeripheral->pvHalContext )->eIrqNum, 1, 0 );
+            HAL_NVIC_EnableIRQ( ( IRQn_Type ) ( ( STM32_UART_HalContext_t * ) pxUartPeripheral->pvHalContext )->eIrqNum );
+
+            if( HAL_UART_Transmit_IT( pxHuart, pvBuffer, ( uint16_t ) xBytes ) != HAL_OK )
+            {
+                lError = IOT_UART_WRITE_FAILED;
+            }
+        }
+    }
+
+    return lError;
+}
+/*-----------------------------------------------------------*/
+
+int32_t iot_uart_read_sync( IotUARTHandle_t const pxUartPeripheral,
+                            uint8_t * const pvBuffer,
+                            size_t xBytes )
+{
+    int32_t lError = IOT_UART_SUCCESS;
+
+    lError = iot_uart_read_async( pxUartPeripheral, pvBuffer, xBytes );
+
+    if( lError == IOT_UART_SUCCESS )
+    {
+        if( xSemaphoreTake( pxUartPeripheral->xSemphr, pdMS_TO_TICKS( IOT_UART_BLOCKING_TIMEOUT ) ) == pdFALSE )
+        {
+            UART_HandleTypeDef * pxHuart = ( UART_HandleTypeDef * ) ( ( STM32_UART_HalContext_t * ) pxUartPeripheral->pvHalContext )->pxHuart;
+            HAL_UART_Abort( pxHuart );
+        }
+    }
+
+    return lError;
+}
+/*-----------------------------------------------------------*/
+
+int32_t iot_uart_write_sync( IotUARTHandle_t const pxUartPeripheral,
+                             uint8_t * const pvBuffer,
+                             size_t xBytes )
+{
+    int32_t lError = IOT_UART_SUCCESS;
+    HAL_StatusTypeDef status = HAL_OK;
+
+    if( ( pvBuffer == NULL ) || ( xBytes == 0 ) || ( pxUartPeripheral == NULL ) || ( pxUartPeripheral->sOpened == IOT_UART_CLOSED ) )
+    {
+        lError = IOT_UART_INVALID_VALUE;
+    }
+    else
+    {
+        UART_HandleTypeDef * pxHuart = ( UART_HandleTypeDef * ) ( ( STM32_UART_HalContext_t * ) pxUartPeripheral->pvHalContext )->pxHuart;
+
+        if( HAL_UART_GetState( pxHuart ) == HAL_UART_STATE_BUSY_TX )
+        {
+            lError = IOT_UART_BUSY;
+        }
+        else
+        {
+            status = HAL_UART_Transmit( pxHuart, pvBuffer, ( uint16_t ) xBytes, IOT_UART_BLOCKING_TIMEOUT );
+
+            if( status != HAL_OK )
+            {
+                lError = IOT_UART_WRITE_FAILED;
+            }
+        }
+    }
+
+    return lError;
+}
+/*-----------------------------------------------------------*/
+
+int32_t iot_uart_close( IotUARTHandle_t const pxUartPeripheral )
+{
+    int32_t lError = IOT_UART_SUCCESS;
+
+    if( ( pxUartPeripheral == NULL ) || ( pxUartPeripheral->sOpened != IOT_UART_OPENED ) )
+    {
+        lError = IOT_UART_INVALID_VALUE;
+    }
+    else
+    {
+        /*lint !e9079 !e9087 handle passed should be of UART_HandleTypeDef *. */
+        UART_HandleTypeDef * pxHuart = ( UART_HandleTypeDef * ) ( ( STM32_UART_HalContext_t * ) pxUartPeripheral->pvHalContext )->pxHuart;
+
+        /* Abort always returns HAL_OK */
```

Comment:
What if the underlying driver implementation changes that behavior at some point?

Reply:
Then our code might break in failed case. What do you suggest in such case? Should we handle the error code any way to be future-proof?

Reply:
Yes, that would seem to be a safer approach. Also, I noticed that we'll return IOT_UART_SUCCESS even if the HAL_UART_DeInit call fails. We should return an error status in the latter case as well.